DHT Request pipelining

vyzo opened the pull request:
The big lock, shared by SendMessage and SendRequest, restricts our ability to issue pipelined requests with potentially significant performance impact -- #88 

This patch implants a goroutine pump for serializing message reads out of line. This allows us to lock only for the duration of the write, which pipelines concurrent requests.

vyzo commented:
Summoning @whyrusleeping @Stebalien

vyzo commented:
Note that there is some subtlety in the fallback to the single request per stream protocol: the single message counter is only incremented in the case of a successful request.

In the case of pipelined requests failing in the read, some of them (if there is more than 2) may fail in their retries due to the old protocol while we are collecting enough samples to fallback.
We can accelerate the convergence by having read errors in retry increment the single message counter.

vyzo commented:
Rebasing on master for #93

Stebalien commented on dht_net.go:
```diff
@@ -195,7 +203,11 @@ func (ms *messageSender) prep() error {
-	ms.r = ggio.NewDelimitedReader(nstr, inet.MessageSizeMax)
+	r := ggio.NewDelimitedReader(nstr, inet.MessageSizeMax)
+	rch := make(chan chan requestResult, requestResultBuffer)
+	go messageReceiver(ms.dht.ctx, rch, r)
```
We could save on long-running go routines by starting these only as needed and shutting them down when we have no more outstanding replies (i.e., by having some form of outstanding reply counter). If we don't keep these around, we could also afford to spin up a second goroutine to manage the outstanding requests.

Stebalien replied:
In general, I'm not a fan of pipelining but out-of-order replies would require a protocol change.

vyzo replied:
Lazily spinning the goroutine is perhaps not so hard to implement, but shutting it down gets tricky -- not sure its worth the complexity. Let me think about it.

vyzo commented on dht_net.go:
```diff
@@ -221,49 +243,164 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+		select {
+		case res = <-resch:
+
+		case <-t.C:
```
good point, will do.

vyzo replied:
we do lose the ability to distinguish ErrReadTimeount however.

Stebalien replied:
You should be able to learn why this failed from `ctx.Err()`.

vyzo replied:
right, that's what I return now -- it's just not `ErrReadTimeout` anymore but rather "context deadline exceeded".

vyzo commented on dht_net.go:
```diff
@@ -221,49 +243,164 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+	defer s.Close()
+
+	w := ggio.NewDelimitedWriter(s)
+	return w.WriteMsg(pmes)
```
sure, small thing to fix.

whyrusleeping commented on dht_net.go:
```diff
@@ -230,31 +247,36 @@ func (ms *messageSender) SendMessage(ctx context.Context, pmes *pb.Message) erro
 func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb.Message, error) {
-	ms.lk.Lock()
-	defer ms.lk.Unlock()
+	defer log.EventBegin(ctx, "dhtSendRequest", ms.dht.self, ms.p, pmes).Done()
```
dont technically need to prefix the event type with "dht", the `log` object is created as a logger for the dht subsystem

vyzo replied:
ipfs log only gives you the event name in the `event` field, so that's kind of necessary to disambiguate events.

vyzo replied:
i mean for practical purposes with grep, otherwise the system field does it just fine.

vyzo replied:
I will remove the prefix, the canonical way to process the event log is jq; and grep becomes only slightly more complicated, it can be a double grep for dht and then SendRequest

whyrusleeping commented on dht_net.go:
```diff
@@ -266,49 +288,166 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+		case res = <-resch:
+
+		case <-rctx.Done():
+			return nil, rctx.Err()
```
If the context is cancelled, do we want to kill off the stream? or is that handled elsewhere?

vyzo replied:
This is the specific request context, so it's inappropriate to kill for the entire stream.

vyzo replied:
Thinking a bit more about this, I think we do want to kill the whole stream after all.
The issue is that if we start having slow responses (over 1m), the pipeline will fill with everything timing out and become unusable until the responses are received (regardless of whether we have stopped waiting).

vyzo replied:
Will implement with a "reset" directive to the message receiver.

whyrusleeping commented on dht_net.go:
```diff
@@ -266,49 +288,166 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+		case next, ok = <-rch:
+			if !ok {
+				return
+			}
```
This might be a little cleaner if we move the logic at the bottom of the loop into this case. Would save us from having to pre-declare those variables

vyzo replied:
ok, was just trying to avoid excessive indentation.

whyrusleeping commented:
This tenatively looks good to me, I'd like to see some tests added that exercise some of the different scenarios (single vs pipelined, slow handlers).

How complicated this is getting makes me think we should make a new DHT protocol handler (version bump) that has message IDs in the protobuf (which should simplify this significantly). Then in the future when we have a 1.0 release, we can drop the old code. Really, this should have been just using a message based interface from the get-go.

cc @Stebalien

vyzo commented:
I will add some more tests as these are important cases we want to make sure we handle right. I am testing with a live node for now.

Re: message ids: Yes, that would make the pump goroutine completely unnecessary and let us handle it with just two locks (read and write) and a queue (for messages read out of order).

vyzo commented:
Some analysis of performance in concurrent requests:

- go-ipfs master, without pipelining:
```
{"event":"findPeerSingleBegin","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T08:32:30.917943022Z"}
{"event":"findPeerSingleBegin","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T08:32:30.918892911Z"}
{"event":"findPeerSingleBegin","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T08:32:30.920377591Z"}
{"duration":157141601,"event":"findPeerSingle","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T08:32:31.076695602Z"}
{"duration":313461714,"event":"findPeerSingle","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T08:32:31.231225466Z"}
{"duration":509875531,"event":"findPeerSingle","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T08:32:31.428626557Z"}
```

- with request pipelining:
```
{"event":"findPeerSingleBegin","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T14:54:58.09138292Z"}
{"event":"findPeerSingleBegin","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T14:54:58.091476318Z"}
{"event":"findPeerSingleBegin","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T14:54:58.09169362Z"}
{"duration":107551379,"event":"findPeerSingle","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T14:54:58.199282563Z"}
{"duration":175856230,"event":"findPeerSingle","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T14:54:58.267292144Z"}
{"duration":256608137,"event":"findPeerSingle","peerID":"QmdFJrz6FXQU15UVCV1Evs4Ftq8xVvAtxfacC23cXySpQZ","system":"dht","time":"2017-09-21T14:54:58.348002273Z"}
```

whyrusleeping commented:
@vyzo update here?

vyzo commented:
will get back to it soon.

whyrusleeping commented:
whoops, sorry.

Stebalien commented:
@vyzo 🙏

vyzo commented:
testing the pipelining effect is difficult -- it's inherently a time-based operation.

Stebalien commented:
@vyzo can we not do a glass-box test that fires repeated requests at a single node? Basically, we just want to make sure that request/responses don't get mixed up and nothing panics.

vyzo commented:
ok, we can try that -- it's simpler to administer.

Stebalien commented:
Looking at some goroutine dumps, this will *really* help on provide. We build up *thousands* of goroutines blocking on sending DHT requests to peers.

vyzo commented:
yeah, I think this will undeniably help. I'll write some more tests.

Stebalien commented:
![profile](https://user-images.githubusercontent.com/310393/34150465-565a4f0a-e4a0-11e7-82b1-a89ac4b73f3e.png)

vyzo commented:
@Stebalien I added a couple of reasonable concurrent request tests.

vyzo commented:
Rebased on master

whyrusleeping commented:
Alright, This LGTM. At some point we need to rework the dht codebase and add a bunch more tests all over.

cc @Stebalien

Stebalien commented on dht_net.go:
```diff
@@ -266,49 +288,165 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+// only a single reset between them.
+// Mutex must be locked.
+func (ms *messageSender) resetSoft(rcount int) {
+	if rcount != ms.rcount {
```
Shouldn't the message receiver be resetting the stream (it observed the error) instead of having the client reset? That would make the most sense, IMO. I'm *really* not a fan of using `rcount` this way.

vyzo replied:
Hrmmmm. We do need a mechanism to unblock the receivers but reset exactly once.

Stebalien replied:
But why can't we do that from the `messageReceiver` loop? That is, when we read an error, reset the stream and return the error.

vyzo replied:
Ok, I think I can rewrite to have the receiver close the stream and potentially eliminate the rcount contraption.

vyzo replied:
Unfortunately If I have the client resetting, there is deadlock potential because of pipeline stalls.

vyzo replied:
Rewrote the reset logic to eliminate the rcount contraption.

vyzo commented on dht_net.go:
```diff
@@ -208,17 +220,22 @@ func (ms *messageSender) prep() error {
 	retry := false
 	for {
+		if ms.singleMes > streamReuseTries {
+			// TODO do this without holding the lock
```
Sure, it deserves one.

vyzo replied:
I think I will just fix it, it's only a matter of removing the defered unlock and moving it to the return sites.

vyzo commented:
@whyrusleeping take another look, I added a stream reset on timeouts (and also addressed some other tidbits).
@Stebalien I have removed the rcount contraption and moved the reset for read errors inside the message receiver.

vyzo commented on dht_net.go:
```diff
@@ -266,49 +301,187 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+
+		case <-rctx.Done():
+			// A read timeout will cause the entire pipeline to time out.
+			// So signal for a stream reset to avoid clogging subsequent requests.
```
the rationale is that if the first request has timed out, then the subsequent requests will likely also time out.

vyzo replied:
also, more requests might be added in the pipeline in the meantime (while the first one has timed out), potentially leading to an avalanche of timeouts; i think it's better to just flush the pipeline and start over.

vyzo commented on dht_net.go:
```diff
@@ -266,49 +301,187 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+	}
+
+	// reset once; needs to happen in a goroutine to avoid deadlock
+	// in case of pipeline stalls
```
i think you are right, we should track which stream we are resetting -- previously, this was protected by the rcount contraption.

vyzo replied:
I added a guarded reset version, `resetStream` which will only stream if it's still the same stream.

vyzo commented on dht_net.go:
```diff
@@ -266,49 +301,187 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+		if err != nil {
+			return nil, err
+		}
+	case <-rctx.Done():
```
probably, although it should happen lazily.

vyzo replied:
Added a reset.

vyzo commented:
@Stebalien do you want to do another round of review here?

whyrusleeping commented:
@vyzo what would make me feel better about merging this, is having a test that tests delayed responses. We could do this by using a datastore with delays (https://github.com/ipfs/go-datastore/blob/master/delayed/delayed.go) and asserting that the requests in aggregate took below a certain amount of time.

vyzo commented:
that sounds like a good idea.

Stebalien commented:
Note: the failing `TestDelayedConcurrentRequest` test is not a bug as we've added an additional datastore read (to avoid overwriting newer records with older records). As we're using a *delayed* datastore, this read significantly slows down the test. The simplest fix is to just bump the expected test time to ~30s.

laser commented:
@Stebalien (cc @whyrusleeping) - I replaced the delayed datastore with a `Mocknet` configured with 100ms of latency (and bumped the number of DHT nodes from 30 to 50). With this configuration, the non-pipelined DHT implementation takes at least twice as long to complete the GetValue + PutValue calls than the pipelined version.

Is there anything else you'd both like to see in this PR before merging?

mishmosh commented:
bump @Stebalien @bigs. When will you have time to review?

bigs commented:
@mishmosh i've got time now! going to look.

Stebalien commented:
@mishmosh I'd like to get some discussion on https://github.com/libp2p/go-libp2p-kad-dht/pull/167 first. In theory, we shouldn't need *any* of this logic.

Stebalien commented:
Status: I've closed #167 because, due to unrelated protocol inefficiencies, it has a lot of bandwidth overhead. When we fix those, we can revisit that PR and delete a *bunch* of code but, until then, we'll stick with one.

whyrusleeping commented:
The right solution there IMO is to push this pipelining logic down into libp2p proper, and expose it as a message/packet based interface. That way we only have to get it right once. It also paves the way for future packet interfaces (cc @lgierth )

Stebalien commented:
So, I agree from a usability perspective. Having a transport-level message system will allow for nice optimizations e.g., when using packet transports.

From a performance perspective, we should be able to just use one stream per message once we improve multistream. For, e.g., QUIC (and yamux), there is no overhead for creating/closing a stream. You can actually send a single packet that opens a stream, has some data, and closes the stream (and the close/open are literally just bits set in the header).

Basically, whatever we do, we'll have to handle protocol muxing and we need to make protocol muxing fast.

laser commented:
@bigs - Is this PR still alive? If so: What's preventing it from being merged?

bigs commented:
@laser was actually just talking about this. essentially no one picked up the ball on this one... it's quite close so i'll make the stylistic changes and we should be ogod

bigs commented:
heya @vyzo, if you've got some spare cycles, it would be great to have you on this merge! the rebase has proven to be pretty challenging, especially since so much has happened since this branch started. feel free to ping me outta band!

vyzo commented:
@bigs feel free to go forward with this, it's been sitting for a long time.

vyzo commented:
I took another look in the code, and complex as it is it still looks ok to me.

bigs commented:
awesome! i believe there is another branch with an attempted rebase. this
hasn’t had the final rebase pushed. let me find it!

On Wed, Aug 22, 2018 at 04:47 vyzo <notifications@github.com> wrote:

> I took another look in the code, and complex as it is it still looks ok to
> me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libp2p/go-libp2p-kad-dht/pull/92#issuecomment-414959665>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANBWvXrmtQ4PEe2b_qD-roHpUrz0F_sks5uTRqIgaJpZM4PaFy0>
> .
>

laser commented:
> i believe there is another branch with an attempted rebase. 

@bigs Indeed there is: `feat/request-pipeline-rebased`

bigs commented:
thanks! let me know if you need another set of eyes on something.

On Fri, Aug 31, 2018 at 2:25 PM, vyzo <notifications@github.com> wrote:

> uh oh, i guess i could.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libp2p/go-libp2p-kad-dht/pull/92#issuecomment-417751354>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANBWkzwg3J9BWvn49myOkVSjrMCcEb7ks5uWX-IgaJpZM4PaFy0>
> .
>

vyzo commented:
So i have rebased this, but I don't know what this business with the delayed tests is.
Another artifact of the rebase is that the bufferedDelimtedWriter use has disappeared -- i don't know what this business is either; this will need to be repatched in.

bigs commented:
@vyzo yeah, that's the same situation i've run into. anything i can assist with?

vyzo commented:
@bigs can you repatch in the buffered writer stuff? Seems like this is important.

Stebalien commented:
> So i have rebased this, but I don't know what this business with the delayed tests is.

@laser was trying to test network latency. The first attempt, using delayed blockstores, wasn't really sufficient. Later tests used the mocknet but I don't think they ever really removed the delayed datastore stuff (we can probably get rid of that).

raulk commented:
@anacrolix – were you intending to review this PR? We need to merge master into it. The conflicts don't look too bad.

anacrolix commented:
I'll try to get on to this now.

anacrolix commented:
I've been poking around this code in master. Why don't we just send a single request per stream?

whyrusleeping commented:
@anacrolix because opening a new stream is expensive right now. Once we get multistream 2, we can just do that.

Stebalien commented:
See: https://github.com/libp2p/go-libp2p-kad-dht/pull/167

anacrolix commented:
I am experimenting with an alternative to this, that optimistically reuses streams, and creates new ones if they're blocked. I believe it won't suffer from pipelining issues, and cross-polluting timeouts etc.

Stebalien commented:
The DHT code has moved on and this would need to be re-implemented.
